File: kotlin-node/karakum/plugins/convertGetPeerCertificate.js
import ts from "typescript";

export default function (node, context, render) {
    if (
        ts.isMethodDeclaration(node)
        && ts.isIdentifier(node.name)
        && node.name.text === "getPeerCertificate"

        && node.parent
        && ts.isClassDeclaration(node.parent)
        && node.parent.name
        && node.parent.name.text === "TLSSocket"

        && node.parameters.length === 1
        && node.parameters[0]
        && node.parameters[0].type
        && ts.isLiteralTypeNode(node.parameters[0].type)
        && (
            node.parameters[0].type.literal.kind === ts.SyntaxKind.TrueKeyword
            || node.parameters[0].type.literal.kind === ts.SyntaxKind.FalseKeyword
        )
    ) {
        return ""
    }

    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertGetPeerCertificate.kt
package node.karakum.plugins

import node.karakum.util.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.SyntaxKind
import typescript.asArray
import typescript.isClassDeclaration
import typescript.isIdentifier
import typescript.isLiteralTypeNode
import typescript.isMethodDeclaration

val convertGetPeerCertificate = createPlugin { node, context, _ ->
    nullable {
        ensure(isMethodDeclaration(node))

        val methodName = node.name
        ensure(isIdentifier(methodName))
        ensure(methodName.text == "getPeerCertificate")

        val classNode = ensureNotNull(node.getParentOrNull())
        ensure(isClassDeclaration(classNode))

        val className = ensureNotNull(classNode.name)
        ensure(className.text === "TLSSocket")

        ensure(node.parameters.asArray().size == 1)

        val firstParameterType = ensureNotNull(
            node.parameters.asArray().getOrNull(0)?.type
        )
        ensure(isLiteralTypeNode(firstParameterType))
        ensure(
            firstParameterType.literal.kind == SyntaxKind.TrueKeyword
                    || firstParameterType.literal.kind == SyntaxKind.FalseKeyword
        )

        ""
    }
}
